Fix f1 scores by correcting calculation of false positives

jonaraphael opened the pull request:
• Fix treatment of false positives, and thereby fix F1 scores
• Add the ability to specify different interpolation during "resize and pad" 
• Genericize the CM to use a class list rather than hardcoded class names

jonaraphael committed:
• Fix treatment of false positives, and thereby fix F1 scores
• Add the ability to specify different interpolation during "resize and pad"
• Genericize the CM to use a class list rather than hardcoded class names

jonaraphael commented:
@rbavery , any thoughts?

rbavery commented on icevision/metrics/confusion_matrix/confusion_matrix_utils.py:
```diff
@@ -36,6 +36,9 @@ def match_records(
     pairs_indices = torch.nonzero(iou_table > iou_threshold)
 
+    # if a prediction has no target ids, then it is background
+    false_positive_indices = torch.nonzero(tensor([0 if any(row) else 1 for row in iou_table > iou_threshold ]))
```
If I understand this correctly, each sample has it's own IOU_table and each row of that table represents a prediction, with elements being the iou of that prediction with a groundtruth. should the iou thresholding condition be applied on the row basis instead of the table basis? otherwise, it seems like iou thresholds for other predictions could impact if another prediction is considered false positive or not?

rbavery replied:
I ran into some trouble trying to empirically test our this code on the provided model

```
load checkpoint from local path: /root/data/experiments/cv2/test_for_ryan/state_dict_model.pt
The model and loaded state dict do not match exactly

size mismatch for backbone.body.layer1.0.conv1.weight: copying a param with shape torch.Size([64, 64, 1, 1]) from checkpoint, the shape in current model is torch.Size([64, 64, 3, 3]).
size mismatch for backbone.body.layer1.1.conv1.weight: copying a param with shape torch.Size([64, 256, 1, 1]) from checkpoint, the shape in current model is torch.Size([64, 64, 3, 3]).
size mismatch for backbone.body.layer1.2.conv1.weight: copying a param with shape torch.Size([64, 256, 1, 1]) from checkpoint, the shape in current model is torch.Size([64, 64, 3, 3]).
size mismatch for backbone.body.layer2.0.conv1.weight: copying a param with shape torch.Size([128, 256, 1, 1]) from checkpoint, the shape in current model is torch.Size([128, 64, 3, 3]).
```

I only ever tested models using the code block in evaluation.ipynb that calls the same CM code as the trainer. so I think the model checkpoint loader from icevision can't handle the progressive resized model checkpoint

jonaraphael committed:
undo class_list import
comment out sahi

jonaraphael committed:
MUST CONNECT WITH 9d0e3e3 OR LATER
Modularize the mask_nms

jonaraphael committed:
BROKEN
Intermediate solution while refactoring the predict and eval pipelines
Roughly paired with [d37fdcd] on cerulean-ml

jonaraphael committed:
WORKING
Aligned to use the same functions as cerulean-ml evaluation.py

jonaraphael committed:
Slight changes to polish confusion matrices
